src/tasks/gitea/gitea-drone-oauth.ts

```diff
@@ -0,0 +1,225 @@
+  } else if (oauthApps.length == 0) {
+    console.log("Gitea doesn't have any oauth application defined")
+  } else if (!oauthApps.some((e) => e.clientId === Buffer.from(remoteSecret.clientId, 'base64').toString())) {
+    console.log('OAuth data did not match with expected secret')
+  } else {
+    return true
+  }
+  return false
+}
+
+async function main() {
+  const giteaUrl: string = env.GITEA_URL.endsWith('/') ? env.GITEA_URL.slice(0, 1) : env.GITEA_URL
+  const droneLoginUrl: string = giteaUrl.replace('gitea', 'drone') + '/login'
+
+  const kc = new k8s.KubeConfig()
+  kc.loadFromDefault()
+  const apiClient = kc.makeApiClient(k8s.CoreV1Api)
+
+  const user = new UserApi(env.GITEA_USER, env.GITEA_PASSWORD, `${giteaUrl}/api/v1`)
+  const oauthOpts = new CreateOAuth2ApplicationOptions()
+  oauthOpts.name = 'otomi-drone'
+  oauthOpts.redirectUris = [droneLoginUrl]
+  const remoteSecret: DroneSecret = await secretExists(apiClient)
+  const oauthApps: OAuth2Application[] = (await user.userGetOauth2Application()).body.filter(
+    (x) => x.name === oauthOpts.name,
+  )
+  // If secret exists (correctly) and oauth apps are (still) defined, everything is good
+  if (isSecretValid(remoteSecret, oauthApps)) {
+    console.log('Gitea Drone OAuth secret exists')
+    const oauthData: DroneSecret = {
+      clientId: Buffer.from(remoteSecret.clientId, 'base64').toString(),
+      clientSecret: Buffer.from(remoteSecret.clientSecret, 'base64').toString(),
+    }
+    await authorizeOAuthApp(giteaUrl, droneLoginUrl, oauthData)
+    return
+  }
+
+  // Otherwise, clear (if necessary)
+  try {
+    await apiClient.deleteNamespacedSecret(k8sSecretName, k8sNamespace)
+  } catch (e) {
+    // Secret didn't exist in
+  }
+  if (oauthApps.length > 0) {
+    for (const oauthApp of oauthApps) {
+      user.userDeleteOAuth2Application(oauthApp.id)
+    }
+  }
+
+  const result: OAuth2Application = (await user.userCreateOAuth2Application(oauthOpts)).body
+  console.log('OAuth app has been created')
+  const oauthData: DroneSecret = {
+    clientId: result.clientId,
+    clientSecret: result.clientSecret,
+  }
+
+  await authorizeOAuthApp(giteaUrl, droneLoginUrl, oauthData)
+  console.log('OAuth app has been authorized')
+
+  const secret: k8s.V1Secret = new k8s.V1Secret()
+  secret.metadata = new k8s.V1ObjectMeta()
+  secret.metadata.name = k8sSecretName
+  secret.metadata.namespace = k8sNamespace
+  secret.data = {
+    clientId: Buffer.from(result.clientId).toString('base64'),
+    clientSecret: Buffer.from(result.clientSecret).toString('base64'),
+  }
+
+  await apiClient.createNamespacedSecret(k8sNamespace, secret)
+  console.log(`New secret ${k8sSecretName} has been created in the namespace ${k8sNamespace}`)
+}
+
+// Run main only on execution, not on import (like tests)
+if (typeof require !== 'undefined' && require.main === module) {
+  main().catch((err) => {
+    if (err instanceof GiteaDroneError) {
+      console.log(err)
```

Comment:
why catch and throw this? Shouldn't the task fail?

Reply:
Because GiteaDroneErrors are not necessarily fatal, it is to indicate that authorization was probably already granted, or that a cookie was not found.
If Authorization was already granted, we don't need to exit fatally out of the application, as that is a state that is fine.